### ISSUE TITLE
unable to start vdi_jupyter.py

### ISSUE THREAD

**Issue by hakaseh** [coecms/nci_scripts#25]
I cannot start vdi_jupyter.py as I get the following error. Could you please help me out?

(jupyter_vdi) DOM-IMA-11GJGH7:~ hakaseh$ python nci_scripts/vdi_jupyter.py 
[09:08:08 jupyter_vdi.py] Using config file: /Users/hakaseh/cosima_cookbook.conf
[09:08:08 jupyter_vdi.py] Checking SSH keys to VDI are configured...
password: 
[09:08:11 jupyter_vdi.py] SSH keys configured OK
[09:08:11 jupyter_vdi.py] Determine if VDI session is already running...
[09:08:13 jupyter_vdi.py] Time remaining: 167:52:41
[09:08:13 jupyter_vdi.py] Determine jobid for VDI session...3555
[09:08:13 jupyter_vdi.py] Get exechost for VDI session...
[09:08:14 jupyter_vdi.py] exechost: vdi-n11.nci.org.au
[09:08:14 jupyter_vdi.py] Running Jupyter on VDI...
[09:08:14 jupyter_vdi.py] Waiting for Jupyter to start...
Traceback (most recent call last):
  File "/Users/hakaseh/nci_scripts/vdi_jupyter.py", line 267, in <module>
    main_argv()
  File "/Users/hakaseh/nci_scripts/vdi_jupyter.py", line 263, in main_argv
    main(args)
  File "/Users/hakaseh/nci_scripts/vdi_jupyter.py", line 239, in main
    ret = s.expect('http://\S*:(?P<jupyterport>\d+)/\?token=(?P<token>[a-zA-Z0-9]+)')
  File "/Users/hakaseh/opt/anaconda3/envs/jupyter_vdi/lib/python3.9/site-packages/pexpect/spawnbase.py", line 343, in expect
    return self.expect_list(compiled_pattern_list,
  File "/Users/hakaseh/opt/anaconda3/envs/jupyter_vdi/lib/python3.9/site-packages/pexpect/spawnbase.py", line 372, in expect_list
    return exp.expect_loop(timeout)
  File "/Users/hakaseh/opt/anaconda3/envs/jupyter_vdi/lib/python3.9/site-packages/pexpect/expect.py", line 179, in expect_loop
    return self.eof(e)
  File "/Users/hakaseh/opt/anaconda3/envs/jupyter_vdi/lib/python3.9/site-packages/pexpect/expect.py", line 122, in eof
    raise exc
pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
<pexpect.pty_spawn.spawn object at 0x100de33a0>
command: /usr/bin/ssh
args: ['/usr/bin/ssh', '-x', '-l', 'hh0162', 'vdi-n11.nci.org.au', '-t', 'bash -l -c "module use /g/data3/hh5/public/modules                  && module load conda/analysis3-unstable                && jupyter notebook --no-browser --port 8889"']
buffer (last 100 chars): b''
before (last 100 chars): b'T: Might try "module unload conda" first.\x1b[m\r\n\r\x1b[K\x1b[?1l\x1b>Connection to vdi-n11.nci.org.au closed.\r\r\n'
after: <class 'pexpect.exceptions.EOF'>
match: None
match_index: None
exitstatus: None
flag_eof: True
pid: 3372
child_fd: 5
closed: False
timeout: 30
delimiter: <class 'pexpect.exceptions.EOF'>
logfile: None
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
searcher: searcher_re:
    0: re.compile(b'http://\\S*:(?P<jupyterport>\\d+)/\\?token=(?P<token>[a-zA-Z0-9]+)')

**Comment by ScottWales** [comment]
Hi Hakase,

What is the output of
```
/usr/bin/ssh -x -l hh0162 vdi-n11.nci.org.au -t bash -l -c "module use /g/data3/hh5/public/modules && module load conda/analysis3-unstable && jupyter notebook --no-browser --port 8889"
```

**Comment by aidanheerdegen** [comment]
This message:
```
Might try "module unload conda" first
```
implies you already have `conda` loaded. Do you have any `module load` statements in your `.bashrc`? If so, you should remove them, and instead create an alias or script to run to load your most commonly used modules after you login.

**Comment by hakaseh** [comment]
> Hi Hakase,
> 
> What is the output of
> 
> ```
> /usr/bin/ssh -x -l hh0162 vdi-n11.nci.org.au -t bash -l -c "module use /g/data3/hh5/public/modules && module load conda/analysis3-unstable && jupyter notebook --no-browser --port 8889"
> ```

(jupyter_vdi) DOM-IMA-11GJGH7:~ hakaseh$ /usr/bin/ssh -x -l hh0162 vdi-n11.nci.org.au -t bash -l -c "module use /g/data3/hh5/public/modules && module load conda/analysis3-unstable && jupyter notebook --no-browser --port 8889"
hh0162@vdi-n11.nci.org.au's password: 
Access denied: user hh0162 (uid=13872) has no active jobs on this node.
Connection closed by 130.56.244.42 port 22



**Currently I am running a VDI session using Strudel on vdi-n19, so I also tried:**

(jupyter_vdi) DOM-IMA-11GJGH7:~ hakaseh$ /usr/bin/ssh -x -l hh0162 vdi-n19.nci.org.au -t bash -l -c "module use /g/data3/hh5/public/modules && module load conda/analysis3-unstable && jupyter notebook --no-browser --port 8889"

**Then I got the following output:**

hh0162@vdi-n19.nci.org.au's password: 
ERROR: Unable to locate a modulefile for 'pangeo/2019.12'
Modules Release 4.3.0 (2019-07-26)
Usage: module [options] [command] [args ...]

Loading / Unloading commands:
  add | load      modulefile [...]  Load modulefile(s)
  rm | unload     modulefile [...]  Remove modulefile(s)
  purge                             Unload all loaded modulefiles
  reload | refresh                  Unload then load all loaded modulefiles
  switch | swap   [mod1] mod2       Unload mod1 and load mod2

Listing / Searching commands:
  list            [-t|-l]           List loaded modules
  avail   [-d|-L] [-t|-l] [-S|-C] [--indepth|--no-indepth] [mod ...]
                                    List all or matching available modules
  aliases                           List all module aliases
  whatis          [modulefile ...]  Print whatis information of modulefile(s)
  apropos | keyword | search  str   Search all name and whatis containing str
  is-loaded       [modulefile ...]  Test if any of the modulefile(s) are loaded
  is-avail        modulefile [...]  Is any of the modulefile(s) available
  info-loaded     modulefile        Get full name of matching loaded module(s)

Collection of modules handling commands:
  save            [collection|file] Save current module list to collection
  restore         [collection|file] Restore module list from collection or file
  saverm          [collection]      Remove saved collection
  saveshow        [collection|file] Display information about collection
  savelist        [-t|-l]           List all saved collections
  is-saved        [collection ...]  Test if any of the collection(s) exists

Shell's initialization files handling commands:
  initlist                          List all modules loaded from init file
  initadd         modulefile [...]  Add modulefile to shell init file
  initrm          modulefile [...]  Remove modulefile from shell init file
  initprepend     modulefile [...]  Add to beginning of list in init file
  initswitch      mod1 mod2         Switch mod1 with mod2 from init file
  initclear                         Clear all modulefiles from init file

Environment direct handling commands:


**I'll try @aidanheerdegen's suggestion now.**

**Comment by hakaseh** [comment]
> This message:
> 
> ```
> Might try "module unload conda" first
> ```
> 
> implies you already have `conda` loaded. Do you have any `module load` statements in your `.bashrc`? If so, you should remove them, and instead create an alias or script to run to load your most commonly used modules after you login.

I assume .bashrc in my VDI home directory? I commented out all of my own "module load" commands, terminated my existing VDI session on Strudel, and did "python nci_scripts/vdi_jupyter.py", but it yielded the same error:



(jupyter_vdi) DOM-IMA-11GJGH7:~ hakaseh$ python nci_scripts/vdi_jupyter.py 
[10:24:23 jupyter_vdi.py] Using config file: /Users/hakaseh/cosima_cookbook.conf
[10:24:23 jupyter_vdi.py] Checking SSH keys to VDI are configured...
password: 
[10:24:27 jupyter_vdi.py] SSH keys configured OK
[10:24:27 jupyter_vdi.py] Determine if VDI session is already running...
[10:24:28 jupyter_vdi.py] No VDI session found
[10:24:28 jupyter_vdi.py] Launching a new VDI session...
[10:24:31 jupyter_vdi.py] Determine jobid for VDI session...3574
[10:24:31 jupyter_vdi.py] Get exechost for VDI session...
[10:24:33 jupyter_vdi.py] exechost: vdi-n19.nci.org.au
[10:24:33 jupyter_vdi.py] Running Jupyter on VDI...
[10:24:33 jupyter_vdi.py] Waiting for Jupyter to start...
Traceback (most recent call last):
  File "/Users/hakaseh/nci_scripts/vdi_jupyter.py", line 267, in <module>
    main_argv()
  File "/Users/hakaseh/nci_scripts/vdi_jupyter.py", line 263, in main_argv
    main(args)
  File "/Users/hakaseh/nci_scripts/vdi_jupyter.py", line 239, in main
    ret = s.expect('http://\S*:(?P<jupyterport>\d+)/\?token=(?P<token>[a-zA-Z0-9]+)')
  File "/Users/hakaseh/opt/anaconda3/envs/jupyter_vdi/lib/python3.9/site-packages/pexpect/spawnbase.py", line 343, in expect
    return self.expect_list(compiled_pattern_list,
  File "/Users/hakaseh/opt/anaconda3/envs/jupyter_vdi/lib/python3.9/site-packages/pexpect/spawnbase.py", line 372, in expect_list
    return exp.expect_loop(timeout)
  File "/Users/hakaseh/opt/anaconda3/envs/jupyter_vdi/lib/python3.9/site-packages/pexpect/expect.py", line 179, in expect_loop
    return self.eof(e)
  File "/Users/hakaseh/opt/anaconda3/envs/jupyter_vdi/lib/python3.9/site-packages/pexpect/expect.py", line 122, in eof
    raise exc
pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
<pexpect.pty_spawn.spawn object at 0x10fc5f250>
command: /usr/bin/ssh
args: ['/usr/bin/ssh', '-x', '-l', 'hh0162', 'vdi-n19.nci.org.au', '-t', 'bash -l -c "module use /g/data3/hh5/public/modules                  && module load conda/analysis3-unstable                && jupyter notebook --no-browser --port 8889"']
buffer (last 100 chars): b''
before (last 100 chars): b'T: Might try "module unload conda" first.\x1b[m\r\n\r\x1b[K\x1b[?1l\x1b>Connection to vdi-n19.nci.org.au closed.\r\r\n'
after: <class 'pexpect.exceptions.EOF'>
match: None
match_index: None
exitstatus: None
flag_eof: True
pid: 5803
child_fd: 5
closed: False
timeout: 30
delimiter: <class 'pexpect.exceptions.EOF'>
logfile: None
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
searcher: searcher_re:
    0: re.compile(b'http://\\S*:(?P<jupyterport>\\d+)/\\?token=(?P<token>[a-zA-Z0-9]+)')

**Comment by aidanheerdegen** [comment]
Have you successfully used {{vdi_jupyter.py}} in the past? If so, have you changed anything since then?

Yes I was referring to the `.bashrc` file in your VDI home directory. It isn't a good idea to load modules in the `.bashrc`, so I would recommend not doing this even if it isn't the proximate cause of this issue.

As you are still getting that `"module unload conda" first` message it seems `conda` is still already loaded. You can try logging directly into the VDI session and see if you already have any modules loaded
```
ssh hh0162 @vdi-n19.nci.org.au
module list
```

**Comment by hakaseh** [comment]
Thanks @aidanheerdegen , you are correct that I had conda still loaded, and fixing it solved the problem! Now vdi_jupyter.py works!

I used to use vdi_jupyter.py successfully before the vdi was upgraded earlier this year. After the upgrade, I created new start-up scripts, .bashrc and .bash_profile. I removed "module conda" in the former, but did not realise that I had the same command in the latter as well. I'll avoid using both scripts as suggested in the future.

Thanks @aidanheerdegen and @ScottWales for your quick replies.
